bpf/lib/nat.h

```diff
@@ -538,21 +543,80 @@ snat_v4_process(struct __ctx_buff *ctx, enum nat_dir dir,
 		tuple.dport = l4hdr.dport;
 		tuple.sport = l4hdr.sport;
 		break;
+#ifndef DISABLE_SNAT_ICMP_SUPPORT
 	case IPPROTO_ICMP:
 		if (ctx_load_bytes(ctx, off, &icmphdr, sizeof(icmphdr)) < 0)
 			return DROP_INVALID;
-		if (icmphdr.type != ICMP_ECHO &&
-		    icmphdr.type != ICMP_ECHOREPLY)
-			return DROP_NAT_UNSUPP_PROTO;
-		if (icmphdr.type == ICMP_ECHO) {
+
+		switch (icmphdr.type) {
+		case ICMP_ECHO:
 			tuple.dport = 0;
 			tuple.sport = icmphdr.un.echo.id;
-		} else {
+			break;
+		case ICMP_ECHOREPLY:
 			tuple.dport = icmphdr.un.echo.id;
 			tuple.sport = 0;
 			icmp_echoreply = true;
+			break;
+		case ICMP_DEST_UNREACH:
+			switch (icmphdr.code) {
+			case ICMP_FRAG_NEEDED:
```

Comment:
Do you have a pcap of such a packet? Asking, as I'm wondering whether not only the original packet needs to be rev-SNAT-ed. The top IPv4 hdr might need to be translated too in order for it to reach the sender pod.

Reply:
Is that what you are looking for ?

https://github.com/cilium/cilium/pull/18414#issuecomment-1047107255

Reply:
Thanks.

> From POD point of view:
> ...
> In f6:8e:6d:a1:ac:bb ethertype IPv4 (0x0800), length 592: 192.168.49.1 > 10.244.0.236: ICMP 10.1.0.100 unreachable - need to frag (mtu 800), length 556

Who did the rev-SNAT translation for this packet? I.e., who changed from 192.168.49.2 to 10.244.0.236 the top-level IPv4 hdr dst addr? My understanding that your proposed change is changing only the src addr of the internal (= the original packet).